src/riscv/sysv.S

```diff
@@ -0,0 +1,782 @@
+    j       set_arg7
+    
+set_arg6_float:
+    addi    t1, t1, -2
+    
+    # when its zero, it means its just a word-sized float
+    bne     t1, zero, set_arg6_double
+    flw     fa6, FFI_SIZEOF_ARG_X6(sp) # load argument
+    j       set_arg7
+    
+set_arg6_double:
+    # otherwise it must be a double we're dealing with
+    fld     fa6, FFI_SIZEOF_ARG_X6(sp)
+
+####################
+## SET ARGUMENT 7 ##
+####################
+
+set_arg7:
+    srli    t1, t0, FFI_FLAG_BITS_X7 # Shift to get the bits for this argument
+    andi    t1, t1, 3               # Mask out the bits for this argument
+    
+    # when its zero, it means its just a word-sized int/ptr
+    bne     t1, zero, set_arg7_float 
+    REG_L   a7, FFI_SIZEOF_ARG_X7(sp) # load argument
+    j       call_it
+    
+set_arg7_float:
+    addi    t1, t1, -2
+    
+    # when its zero, it means its just a word-sized float
+    bne     t1, zero, set_arg7_double
+    flw     fa7, FFI_SIZEOF_ARG_X7(sp) # load argument
+    j       call_it
+    
+set_arg7_double:
+    # otherwise it must be a double we're dealing with
+    fld     fa7, FFI_SIZEOF_ARG_X7(sp)
+    
+#else
+
+## START RISCV SOFT-FLOAT LOADING ##
+
+    # In the soft-float case, we have no primitive datatype
+    # that has a size of >8 bytes. Therefore, we can 
+    # just load everything quite easily and nicely.
+
+    REG_L   a0, FFI_SIZEOF_ARG_X0(sp) # load argument
+    REG_L   a1, FFI_SIZEOF_ARG_X1(sp) # load argument
+    REG_L   a2, FFI_SIZEOF_ARG_X2(sp) # load argument
+    REG_L   a3, FFI_SIZEOF_ARG_X3(sp) # load argument
+    REG_L   a4, FFI_SIZEOF_ARG_X4(sp) # load argument
+    REG_L   a5, FFI_SIZEOF_ARG_X5(sp) # load argument
+    REG_L   a6, FFI_SIZEOF_ARG_X6(sp) # load argument
+    REG_L   a7, FFI_SIZEOF_ARG_X7(sp) # load argument
+
+#endif  
+
+call_it:
+    # First, we fix the stack pointer to point to the first argument
+    # passed on the stack.
+    add     sp, sp, FFI_SIZEOF_ARG_X8
+
+    # Load the function pointer
+    REG_L   t4, -FFI_SIZEOF_ARG_X4(fp)
+
+    # When the return value pointer is NULL, assume no return value.
+    REG_L   t1, -FFI_SIZEOF_ARG_X5(fp)
+    beq     t1, zero, return_void
+    
+    # is the return type an int? if not, jump ahead
+    ori     t3, zero, FFI_TYPE_INT
+    bne     t2, t3,   return_int32
+    
+    jalr    t4 # call the function
+    
+    # We reload the return pointer because it was in a temp reg and
+    # there was just a function call.
+    REG_L   t0, -FFI_SIZEOF_ARG_X5(fp)
+    REG_S   a0, 0(t0)
+    j       epilogue
+
+return_int32:
+    ori     t3, zero, FFI_TYPE_SINT32
+#ifndef __riscv_soft_float
+    bne     t2, t3, return_float
+#else
+    bne     t2, t3, return_struct_d_soft
+#endif
+    
+    jalr    t4 # call the function
+    
+    # We reload the return pointer because it was in a temp reg and
+    # there was just a function call.
+    REG_L   t0, -FFI_SIZEOF_ARG_X5(fp)
+    sw      a0, 0(t0)
+    j       epilogue
+    
```

Comment:
Is the call of the actual function being delayed so that you don't have to save the flags value across the call?  Except that you are saving the flags, in slot 3 of the stack frame...

The if-else-if (times 10) structure is less than efficient.  You should be able to create a small computed branch construct with AUIPC to perform this efficiently.  See the aarch64 port for a good example of this.